### PR TITLE
CLI: Generate a new project on init in empty directory

### PR DESCRIPTION
**Pull request by ShaunEvening**
## What I did
- [x] Create exec function in the CLI utilities
- [x] Create function to create a new project in empty directory
- [x] If `storybook init` is run on an empty directory, ask user to select a project to generate.
- [x] Throw a categorized error if the scaffolding fails
- [x] Update documentation for getting started
- [x] Add test cases into the CI

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

1. compile packages with `yarn task --task compile`
2. alias the local version of the CLI in your `.zshrc`
3. Create an empty directory and cd into it
4. Run your aliased local CLI

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [x] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This pull request has been released as version [`0.0.0-pr-24997-sha-f3766dc1`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-24997-sha-f3766dc1). Install it by pinning all your Storybook dependencies to that version.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-24997-sha-f3766dc1`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-24997-sha-f3766dc1) |
| **Triggered by** | @Integrayshaun |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`cli/empty-project-init`](https://github.com/storybookjs/storybook/tree/cli/empty-project-init) |
| **Commit** | [`f3766dc1`](https://github.com/storybookjs/storybook/commit/f3766dc1c86bcf419a8ff00442dae08a68073884) |
| **Datetime** | Tue Dec  5 14:16:27 UTC 2023 (`1701785787`) |
| **Workflow run** | [7101997420](https://github.com/storybookjs/storybook/actions/runs/7101997420) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=24997`_
</details>

### CODE REVIEW

File: code/lib/cli/src/scaffold-new-project.ts [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+      pnpm: 'pnpm create vite@latest --yes . --template react-ts && pnpm i --prefer-offline',
+    },
+  },
+  'nextjs-ts': {
+    displayName: {
+      type: 'Next.js',
+      language: 'TS',
+    },
+    createScript: {
+      npm: 'npm create next-app . -- --typescript --use-npm --eslint --tailwind --app --import-alias="@/*" --src-dir',
```
Comment (shilman) [inline code/lib/cli/src/scaffold-new-project.ts:43]: Do we want `--no-app` here because Storybook doesn't support RSC yet?

**Review by shilman — COMMENTED** [review]
Looking great so far!

File: code/lib/cli/src/scaffold-new-project.ts (hunk 1 above)
Comment (Integrayshaun) [reply code/lib/cli/src/scaffold-new-project.ts]: Good catch! :D updated!

File: code/lib/cli/src/scaffold-new-project.ts [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+  'react-vite-ts': {
+    displayName: {
+      type: 'React',
+      builder: 'Vite',
+      language: 'TS',
+    },
+    createScript: {
+      npm: 'npm create vite@latest --yes . -- --template react-ts',
+      yarn: 'yarn create vite@latest --yes . --template react-ts',
+      pnpm: 'pnpm create vite@latest --yes . --template react-ts && pnpm i --prefer-offline',
```
Comment (valentinpalkovic) [inline code/lib/cli/src/scaffold-new-project.ts:35]: If this runs actually on the user's system, we should ditch the `--prefer-offline` flag. It was just introduced for performance reasons on CI to generate the sandboxes.

File: code/lib/cli/src/utils.ts [hunk 3]
```diff
@@ -130,3 +132,32 @@ export const isCorePackage = (pkg: string) =>
...
+    child.on('exit', (code) => {
+      if (code === 0) {
+        resolve(undefined);
+      } else {
+        logger.error(chalk.red(`An error occurred while executing: \`${command}\``));
+        reject(new Error(`Command failed with exit code: ${code}`));
+      }
+    });
+  });
+};
```
[the remark this answers is not in the document]
Comment (Integrayshaun) [reply code/lib/cli/src/utils.ts:163]: I just followed what I saw around the codebase but happy to switch over to using execa :)

**Comment by shilman** [comment]
@Integrayshaun you asked about how to test? i would probably add N sandboxes to the `daily` run that:
1. create an empty directory
2. sets an environment variable `STORYBOOK_INIT_EMPTY_TYPE` (or something)
3. update the code to not prompt the user if `STORYBOOK_INIT_EMPTY_TYPE` is set

That way the rest of the test infrastructure can run on these sandboxes. It's a little "heavy" but it's basically making sure that the codepath of instantiating a project directly via an external installers & initializing using the CLI works, rather than our main sandbox mechanism, which is a different codepath and can lag behind external installers by a day or more.

WDYT?

**Comment by yannbf** [comment]
I think the canary release section was deleted in the PR template, that's why it won't get updated. Here's the canary for anyone who wants to try it out:
```
npx storybook@0.0.0-pr-24997-sha-6379ff8d init
```

**Comment by valentinpalkovic** [comment]
`yarn dlx storybook@0.0.0-pr-24997-sha-6379ff8d init` fails with the following error message:

```
error: Error: Command failed with exit code 1: yarn create vite@latest --yes . --template react-ts
    error An unexpected error occurred: "Cannot find module 'pnpapi'
    Require stack:
    - /Users/-/.cache/node/corepack/yarn/1.22.19/lib/cli.js
    - /Users/-/.cache/node/corepack/yarn/1.22.19/bin/yarn.js
    Require stack:
    - /Users/-/.cache/node/corepack/yarn/1.22.19/lib/cli.js
    - /Users/-/.cache/node/corepack/yarn/1.22.19/bin/yarn.js".
```

**Comment by yannbf** [comment]
When running:
```
npx storybook@0.0.0-pr-24997-sha-6379ff8d init
```

and choosing angular, when the CLI tries to run Storybook, it fails:

image

The generation works correctly, so if you run Storybook manually later it does work.

**Comment by yannbf** [comment]
When using `npx` and selecting Next.js, when invoking the Next.js CLI it errors given that the directory is not fully empty.

image

There is a `.cache` directory before running the Next.js CLI command, which I believe is generated by Storybook when using telemetry.

**Comment by yannbf** [comment]
The empty directory flow is triggered when there are hidden directories (so the directory is not fully empty):

image

File: code/lib/cli/src/initiate.ts [hunk 4]
```diff
@@ -292,11 +254,10 @@ async function doInitiate(
     pkgMgr = 'npm';
   }
 
-  const cwdFolderEntries = readdirSync(process.cwd());
-  const isEmptyDir =
-    cwdFolderEntries.length === 0 || cwdFolderEntries.every((entry) => entry.startsWith('.'));
+  const packageManager = JsPackageManagerFactory.getPackageManager({
+    force: pkgMgr || inferPackageManagerFromUserAgent(),
```
Comment (shilman) [inline code/lib/cli/src/initiate.ts:258]: If `JsPackageManagerFactory.getPackageManager` is not doing the right thing, why don't we update it to use `inferPackageManagerFromUserAgent` internally instead?

File: code/lib/cli/src/sandbox-templates.ts [hunk 5]
```diff
@@ -528,6 +532,156 @@ const internalTemplates = {
     isInternal: true,
     skipTasks: ['bench'],
   },
+  'internal/npm/empty/react-vite/default-ts': {
```
Comment (shilman) [inline code/lib/cli/src/sandbox-templates.ts:535]: Do we still need these sandboxes now that we have your new CI task above?

Comment (Integrayshaun) [reply code/lib/cli/src/sandbox-templates.ts:535]: We don't need them 😁 I just didn't want to delete all the configs until we were sure we didn't want them

File: code/lib/cli/src/initiate.ts (hunk 4 above)
Comment (Integrayshaun) [reply code/lib/cli/src/initiate.ts:258]: That's a great idea!

**Review by shilman — APPROVED** [review]
Woohooo!!! 🙌

**Pull request merged**